modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/DefaultExtensionManager.java

```diff
@@ -0,0 +1,425 @@
+import org.mule.module.extension.internal.runtime.resolver.ResolverSet;
+import org.mule.module.extension.internal.runtime.resolver.StaticValueResolver;
+import org.mule.module.extension.internal.runtime.resolver.ValueResolver;
+import org.mule.util.ObjectNameHelper;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Default implementation of {@link ExtensionManagerAdapter}
+ *
+ * @since 3.7.0
+ */
+public final class DefaultExtensionManager implements ExtensionManagerAdapter, MuleContextAware, Initialisable
+{
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultExtensionManager.class);
+
+    private final ExtensionRegister register = new ExtensionRegister();
+    private final ServiceRegistry serviceRegistry = new SPIServiceRegistry();
+
+    private MuleContext muleContext;
+    private ObjectNameHelper objectNameHelper;
+    private ExtensionDiscoverer extensionDiscoverer = new DefaultExtensionDiscoverer(new DefaultExtensionFactory(serviceRegistry), serviceRegistry);
+
+    /**
+     * Searches the mule registry for instances of {@link ConfigurationInstanceProvider}
+     * and registers them through the {@link #registerConfigurationInstanceProvider(String, ConfigurationInstanceProvider)}
+     * method
+     */
+    @Override
+    public void initialise() throws InitialisationException
+    {
+        for (Map.Entry<String, ConfigurationInstanceProvider> instanceProviderEntry : muleContext.getRegistry().lookupByType(ConfigurationInstanceProvider.class).entrySet())
+        {
+            ConfigurationInstanceProvider<?> instanceProvider = instanceProviderEntry.getValue();
+            registerConfigurationInstanceProvider(instanceProviderEntry.getKey(), instanceProvider);
+        }
+    }
+
+    @Override
+    public List<Extension> discoverExtensions(ClassLoader classLoader)
+    {
+        LOGGER.info("Starting discovery of extensions");
+
+        List<Extension> discovered = extensionDiscoverer.discover(classLoader);
+        LOGGER.info("Discovered {} extensions", discovered.size());
+
+        ImmutableList.Builder<Extension> accepted = ImmutableList.builder();
+
+        for (Extension extension : discovered)
+        {
+            if (registerExtension(extension))
+            {
+                accepted.add(extension);
+            }
+        }
+
+        return accepted.build();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean registerExtension(Extension extension)
+    {
+        LOGGER.info("Registering extension {} (version {})", extension.getName(), extension.getVersion());
+        final String extensionName = extension.getName();
+
+        if (register.containsExtension(extensionName))
+        {
+            return maybeUpdateExtension(extension, extensionName);
```

Comment:
the use case when this was done was to give the user the ability to change the version of an extension in runtime. In which case, the new version was to be rediscovered and the model updated.

If this no longer makes sense, we can change it, but we should file a specific issue for that defining the behaviour we actually want